modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseNodeService.java

```diff
@@ -492,8 +499,60 @@ void retrieveDatabase(
         });
     }
 
-    public Set<String> getAvailableDatabases() {
-        return Set.copyOf(databases.keySet());
+    public Set<DatabaseInfo> getAvailableDatabases() {
+        /*
+         * We return all databases that exist on the node, including manually-configured databases and downloader databases. There will
+         * not be more than one entry per database name. If more than one exist, only the active one will be returned. Downloader
+         * databases take priority over manually-configured databases.
+         */
+        Map<String, DatabaseInfo> allDatabases = new HashMap<>();
+        for (Map.Entry<String, DatabaseReaderLazyLoader> entry : configDatabases.getConfigDatabases().entrySet()) {
+            DatabaseReaderLazyLoader databaseReaderLazyLoader = entry.getValue();
+            try {
+                final Metadata metadata = databaseReaderLazyLoader.getMetadata();
+                allDatabases.put(
+                    entry.getKey(),
+                    new DatabaseInfo(
+                        entry.getKey(),
+                        "config",
+                        databaseReaderLazyLoader.getArchiveMd5(),
+                        databaseReaderLazyLoader.getMd5(),
+                        metadata.getBuildDate().getTime(),
+                        metadata.getDatabaseType()
+                    )
+                );
+            } catch (FileNotFoundException e) {
+                /*
+                 * Since there is nothing to prevent a database from being deleted while this method is running, it is possible we get an
+                 * exception here because the file no longer exists. We just log it and move on -- it's preferable to synchronization.
+                 */
+                logger.trace(Strings.format("Unable to get metadata for config database %s", entry.getKey()), e);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        for (Map.Entry<String, DatabaseReaderLazyLoader> entry : databases.entrySet()) {
+            DatabaseReaderLazyLoader databaseReaderLazyLoader = entry.getValue();
+            final Metadata metadata;
+            try {
+                metadata = databaseReaderLazyLoader.getMetadata();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            allDatabases.put(
+                entry.getKey(),
+                new DatabaseInfo(
+                    entry.getKey(),
+                    "downloader",
+                    databaseReaderLazyLoader.getArchiveMd5(),
+                    databaseReaderLazyLoader.getMd5(),
+                    metadata.getBuildDate().getTime(),
+                    metadata.getDatabaseType()
+                )
+            );
+        }
+        return new HashSet<>(allDatabases.values());
```

Comment:
A minor nit, but what do you think about `return Set.copyOf(allDatabases.values())` or some-such here instead?